chirp/drivers/baofeng_digital.py

```diff
@@ -0,0 +1,492 @@
+        MemoryChunk(start=0x330, end=0x33F, read_only=True),
+        MemoryChunk(start=0x380, end=0x3DF, read_only=False),
+    ],
+)
+PROTO_B = ProtoVersion(
+    chan_count=16,
+    serial_magic=b'\x80\x82',
+    mem_chunks=PROTO_A.mem_chunks,
+)
+PROTO_C = ProtoVersion(
+    chan_count=16,
+    serial_magic=b'\x20\x22',
+    mem_chunks=PROTO_A.mem_chunks,
+)
+PROTO_D = ProtoVersion(
+    chan_count=30,
+    serial_magic=b'\x80\x82',
+    mem_chunks=[
+        MemoryChunk(start=0x000, end=0x1EF, read_only=False),
+        MemoryChunk(start=0x2B0, end=0x2BF, read_only=False),
+        MemoryChunk(start=0x330, end=0x33F, read_only=True),
+        MemoryChunk(start=0x380, end=0x3DF, read_only=False),
+    ],
+)
+
+
+class BaofengDigital(chirp_common.CloneModeRadio,
+                     chirp_common.ExperimentalRadio):
+    """Base for all digital radios from Fujian Baofeng Electronics Co. Ltd."""
+    VENDOR = "Baofeng"
+    BAUD_RATE = 9600
+
+    MEM_FORMAT = """
+        #seekto 0x0010;
+        struct {{
+            lbcd rx_freq[4];
+            lbcd tx_freq[4];
+            ul16 rx_tone;
+            ul16 tx_tone;
+            u8 hopping:1,
+               unused1:2,
+               scan:1,
+               high_power:1,
+               narrow:1,
+               digital:1,
+               bcl_disabled:1;
+            u8 unused2:3,
+               encryption_key:5;
+            u16 unused3;
+        }} memory[{chan_count}];
+    """
+
+    _PREAMBLE_CMDS = [
+        (b'\x02', b'P3107\xf7\x00\x00'),
+        (CMD_ACK, CMD_ACK),
+        (b'R\x01\x30\x08', b'W\x01\x30\x08\xff\xff\xff\xff\xff\xff\xff\xff'),
+        (CMD_ACK, CMD_ACK),
+    ]
+
+    _WRITE_PREAMBLE_CMDS = [
+        (b'E', b'F'),
+        (b'\x02\x80\x82OGRAM', CMD_ACK),
+        (b'\x02', b'P3107\xf7\x00\x00'),
+        (CMD_ACK, CMD_ACK),
+    ]
+
+    _VOICES_PER_REGION = {
+        0x00: ('English', 'Chinese'),
+        0x01: ('Chinese', 'Russian'),
+        0x02: ('Russian', 'English'),
+        0x03: ('Arabic', 'English'),
+        0x04: ('Arabic', 'Chinese'),
+        0x05: ('Arabic', 'Russian'),
+        0x06: ('Chinese', 'Portuguese'),
+        0x07: ('English', 'Portuguese'),
+        0x08: ('Chinese', 'Spanish'),
+        0x09: ('English', 'Spanish'),
+        0x0A: ('Chinese', 'German'),
+        0x0B: ('English', 'German'),
+        0x0C: ('Chinese', 'French'),
+        0x0D: ('English', 'French'),
+    }
+
+    @property
+    def _memsize(self):
+        return self._proto.mem_chunks[-1].end + 1
+
+    def get_features(self):
+        rf = chirp_common.RadioFeatures()
+        rf.has_rx_dtcs = True
+        rf.has_bank = False
+        rf.has_name = False
+        rf.has_tuning_step = False
+        rf.can_odd_split = True
+        rf.has_cross = True
+        rf.memory_bounds = (1, self._proto.chan_count)
+        rf.valid_bands = self.VALID_BANDS
+        rf.valid_duplexes = ["", "+", "-", "split"]
+        rf.valid_modes = ['FM', 'NFM', 'DIG', 'DN']
```

Comment:
`DN` means Yaesu system fusion, so unless this radio implements that, you may not use `DN` here. Also `DIG` is reserved for the digital setting in HF radios. It's not a good use of the mode, but that's what it has meant in CHIRP in the past, so it needs to remain.

CHIRP needs a consistent view of the world in terms of what modes mean, so that you copy a memory from one radio to another, we can select the proper mode *or* reject it if it's not compatible. For example, if you open an FT3DR image and copy a `DN` memory out of it and paste into this radio, CHIRP should reject that paste because your radio doesn't support actual YSF DN mode (assuming the above).

What digital mode is this actually?

Reply:
I have absolutely no idea. Their sales pitch says it's GMSK, and that is all the information I have.

It supports 25KHz and 12.5KHz using the same modulation, and I don't know any  standard mode that uses that modulation with variable width?

Reply:
Yeah there has to be more than just GMSK. Also I doubt it's 25kHz or 12.5kHz wide, but rather those are the channel spacing recommendations, but who knows for sure I guess.

How about we just support FM,NFM and make the digital mode a .extra toggle for now (i.e. FM+extra['digital'] means "25kHz wide in digital mode" ? Sort of like we do for radios with the "scrambler" modules - we don't consider that a different mode, just a flag on an existing one. If we figure out that it is actually something else (or a second radio starts supporting it) we can add it into the mode list?